### ISSUE TITLE
Issues with sys ontology in SPARQL queries

### ISSUE THREAD

**Issue by FlipGFlopTR** [thomsonreuters/CM-Well#406]
(Dan B suggested I log this as an issue in github as there may be 1 or 2 bugs; I hope this is the correct place to record it.)

I've been trying to write some basic SPARQL sub-graph and graph queries using CM Well (Octopus) and the full pathname for the sys ontology, eg this one:

> SELECT DISTINCT ?pred ?value
> WHERE {
> <http://permid.org/1-100131> ?pred ?value .
> } ORDER BY (?pred)

As a partial graph query this returns results like this one:

> <http://localhost:9000/meta/sys#parent>	"/permid.org"

As a full graph query it returns this:

> <http://cm-well-prod.int.thomsonreuters.com/meta/sys#parent> "/permid.org"

*** 1 - Why is the name of the sys ontology inconsistent? ***            

If I try a modified full graph query:

> prefix sys: <http://cm-well-prod.int.thomsonreuters.com/meta/sys>
> 
> SELECT DISTINCT
> ?pred ?value
> WHERE {
> <http://permid.org/1-100131> sys:parent ?value .
> }
> ORDER BY (?pred)

then this generates this error:

> [Error] Namespace http://cm-well-prod.int.thomsonreuters.com/meta/ does not exist 

*** 2 - why does this generate an error? ***

**Comment by hochgi** [comment]
Seems like there's a `#` missing in the prefix declaration.
you actually tries to search `sysparent` under `http://.../meta/` namespace.

BTW, `sys` (and `nn`) are special namespaces. although it will work the way you wrote it,
It works only because CM-Well uses some heuristics to infer that the domain points to a CM-Well instance.
a better way to define it would be to use the `cmwell` "protocol", e.g:
```
prefix sys: <cmwell://meta/sys#>

… sys:parent …
```
This will always work. and is consistent among different clusters.

**Comment by FlipGFlopTR** [comment]
@hochgi - thanks for the rapid response; unfortunately the tips you suggest don't resolve the issue.

1. I've tried both of the http forms and they both fail with the similar error messages:

prefix sys:   <http://cm-well-prod.int.thomsonreuters.com/meta/sys>

[Error] Namespace http://cm-well-prod.int.thomsonreuters.com/meta/ does not exist

prefix sys:   <http://cm-well-prod.int.thomsonreuters.com/meta/sys#>

[Error] Namespace http://cm-well-prod.int.thomsonreuters.com/meta/sys# does not exist

2.  I tried the cmwell: form and, using the _sparql endpoint the query doesn't return any results (the query times out).

As I'm a beginner with both CM Well and SPARQL, I may be making some basic mistake so I've attached a log of the queries & results using the _sparql endpoint.

[prefix-test.log](https://github.com/thomsonreuters/CM-Well/files/1651444/prefix-test.log)

**Comment by bryaakov** [comment]
@FlipGFlopTR please note that the `_sparql` endpoint is for the **SPARQL on the Whole Graph** feature, which is currently being stabilized and optimized, (and we are thinking of a re-write).
A more robust SPARQL endpoint CM-Well provides is `_sp` which is for the **SPARQL on the Subgraph** feature. Please read [relevant documentation](https://github.com/thomsonreuters/CM-Well/blob/master/docs/API.Query.ApplySPARQLToQueryResults.md) and try to run your queries against `_sp` - if possible (I assume you can distinguish the population of data your are running your query against from the entire knowledge-graph).

**Comment by FlipGFlopTR** [comment]
@bryaakov - thanks.  I actually started with a sub-graph query using the _sp endpoint which I then tried to re-write as a full graph query using the _sparql endpoint.  You can see the discussion I had with @TRnonodename  on the hub here - https://thehub.thomsonreuters.com/message/709447?commentID=709447#comment-708854.

However, there are issues with the 'sys' ontology in the subgraph interface too.  I've tried it with three versions of the prefix statement for the sys ontology (see code below) and the only one which returns valid results is the http form using localhost as the root for the ontology path.  The other two formats, including the cmwell: form complete but return an empty table ie are syntactically valid but semantically incorrect.

**In summary, the cmwell form of the prefix suggested by @hochgi doesn't work properly with either endpoint: for _sp it returns an empty table; for _sparql it times out.  The only version of the http form of the prefix that works is the localhost one when run on the _sp endpoint.**

_The ability to filter queries by namespace in SPARQL against cm-well be vital because of the way the content in the various namespaces like permid.org, data.thomsonreuters.com overlap.  This is why I'm trying to get the sys ontology working in SPARQL queries._

`
PATHS
/permid.org?op=search&qp=type.rdf:Country&with-data&length=1000
/permid.org?op=search&qp=type.rdf:Currency&with-data&length=1000

SPARQL
prefix geo: <http://permid.org/ontology/geography/>
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
prefix sys: <http://localhost:9000/meta/sys#> # works
# prefix sys: <http://cm-well-prod.int.thomsonreuters.com/meta/sys#> # no results
# prefix sys: <cmwell://meta/sys#> # no results

SELECT ?pred (COUNT(?Value) AS ?Count)
WHERE {
  ?infoton ?pred ?Value.
  ?infoton sys:parent "/permid.org". # this is, strictly, redundant as the PATHs clause restricts to the permid.org namespace
  ?infoton rdf:type geo:Country. # if the Currency path was removed then this would be redundant 
}
GROUP BY (?pred)
ORDER BY (?pred)
`

**Comment by TRnonodename** [comment]
As a point of courtesy to the community: let's try not to post links to thehub here since they're not accessible to non TR employees.

**Comment by FlipGFlopTR** [comment]
I did think twice before posting the link but decided that it contained information that could be useful @bryaakov, who would be able to read it but I've removed the link now.

**Comment by bryaakov** [comment]
Just one more question: Why do we need to query against System Fields? Is it only for the parent? It may be a nice-to-have feature, to enable querying system fields, however, it sounds to me as a bad design choice. SPARQL should work against the data itself, rather than counting on CM-Well's meta data. Again, I'm not saying it's impossible, but for this case, as an example, why not use **the subject URI itself** to distinguish, rather than sys:parent? 
e.g.
```FILTER regex(STR(?infoton), "http://permid.org/*")```

**Comment by FlipGFlopTR** [comment]
@bryaakov - in this specific case I need to restrict the query to a specific namespace in CM Well.  When I looked at the infotons then using the sys ontology and specifically the parent predicate was the obvious way to do it as this is the way that the namespace is modeled in the RDF.

The FILTER 'trick' appears to be functionally equivalent (when I tried it with the _sp endpoint it seems to give the same results (to an eye check)).  I call it a trick because it is relying on the syntax of the subject to extract the namespace but that's probably something one argue about.

However, I don't need to do this with the _sp endpoint because the PATHS clause allows me to do this filtering and, if I try to use it with the _sparql endpoint, I run into a couple of issues:

- The CM Well documentation states that 'Using the SPARQL OPTIONAL and FILTER options is very expensive when performed on the entire graph. Avoid using these options if possible.'

-More importantly, if I try to run the query (see below) using the _sparql endpoint then I get an error:

> [Error] Each triple-matching must have binding of a subject, a predicate or an object. If you'd like to download entire CM-Well's content, please use the Stream API

```
prefix geo: <http://permid.org/ontology/geography/>
prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>

SELECT ?pred (COUNT(?Value) AS ?Count)
WHERE {
  ?infoton ?pred ?Value.
  FILTER regex(STR(?infoton), "http://permid.org/*")
  ?infoton rdf:type geo:Country. # if the Currency path was removed then this would be redundant 
}
GROUP BY (?pred)
ORDER BY (?pred)
```

In conclusion: I think that the sys ontology does need to work in CM Well SPARQL but I'm not averse to using the FILTER approach if it let's me solve the problem.  However, the CM Well documentation advises against it and the s/w doesn't seem to want me to!

**Comment by bryaakov** [comment]
Hi Phil, sorry for the delay in responding. 
1. I disagree with calling it a trick. One of the beautiful things in CM-Well is that each triple is "physically" located in the URL of its subject. Having an RDF dataset where the subject are <http://some.domain/a/path/*> will actually be located in CM-Well cluster within the URL cm-well-clusrer/some.domain/a/path - think of it as REST. 
So **logically** this is not a trick at all, it is relying on one of CM-Well core principles. 
2. With regards to **performance** - you're absolutely right. Bottom line: You convinced me to prioritize this bugfix and allow querying system fields. Please stay tuned and wait for our Rhino release.
The error message you got was because the SPARQL Engine detected you have a triple pattern without any constraints: `?infoton ?pred ?Value.` which was translated to `?op=stream&recursive&with-data` equivalent.
Keep in mind we are still working on stabilizing and optimizing the _sparql endpoint, however. One of the things I'd like to add is having FILTERs somehow embedded in the queries the SPARQL Engine invoke to CM-Well internal APIs, rather than filter the data in-memory, which will of course give another performance boost.

**Comment by FlipGFlopTR** [comment]
@bryaakov - thanks.

1. I may have been a harsh when I said 'trick' but as a general principle I think it is better to have an explicit model for a concept like the namespace than require knowledge of the URI structure:.  This principle is built into the CM Well 'native' UI and I was looking for a way to restrict my query to the namespace because the UI inspired me to!

2.  This sounds good: the sys: ontology looks like a normal ontology so we ought to be able to query it.

Clearly the 'closer to the disk' that the FILTER statements are applied the more efficient and better performing it will be (I've worked on systems where simple predicates are checked at the disk level!).

**Comment by AnnaGrebnev** [comment]
fixed links
